{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SigCom LINCS Documentation\n",
    "\n",
    "## Table of contents\n",
    "- About SigCom LINCS\n",
    "- Computing the L1000 Signatures\n",
    "- Benchmarking the L1000 Signatures\n",
    "- Web Interface\n",
    "    - Signature Search\n",
    "    - Performing Metadata Search\n",
    "    - Metadata Dedicated Landing Pages\n",
    "    - The Download Tab\n",
    "    - The API Page\n",
    "- Using the APIs\n",
    "    - Full-text metadata search\n",
    "    - Filtering by metadata fields\n",
    "    - Finding available fields\n",
    "    - Counting the results\n",
    "    - Fetching the value counts\n",
    "    - Performing Signature Search\n",
    "    - Listing available datasets and databases for search\n",
    "- References\n",
    "   \n",
    "\n",
    "## About SigCom LINCS\n",
    "SigCom LINCS enables users to query the datasets and signatures generated by the LINCS consortium. SigCom LINCS is built with the Signature Commons framework (Figure 1). SigCom LINCS has RESTful APIs, and a dedicated web interface for exploring the LINCS data. In this documentation, we provide details on how the LINCS data was processed, ingested, and benchmarked. We also provide tutorials on how to use the website and the APIs.\n",
    "\n",
    "![sigcom_architecture.jpg](img/sigcom-architecture.jpg)\n",
    "**Figure 1** Signature Commons Architecture <a name=\"architecture\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing the L1000 Signatures\n",
    "\n",
    "Level 3 L1000 gene expression profile datasets were downloaded from CLUE.io (https://clue.io/) on June 2, 2021, and divided into batches based on the batch ID of each profile. For example, profiles corresponding to the IDs LJP008_NEU_24H_X1_B21:N15 and LJP008_NEU_24H_X2_B21:O06 would both belong to the batch LJP008_NEU_24H.\n",
    "\n",
    "Using the Level 5 signature metadata file, also downloaded from CLUE.io on June 2, 2021, the replicate profile IDs corresponding to each signature were identified. For each signature, the signature was computed using the Characteristic Direction method [1], which compares the corresponding replicate profiles (the “perturbations”) to all other profiles in the batch to which the signature belongs (the “controls”). The resulting signature is a vector of coefficients representing the differential expression of each gene in the perturbation profiles when compared to the control profiles.\n",
    "\n",
    "All signatures may be accessed individually via the persistent_id field in the metadata for each signature. Compiled signatures by perturbation type can also be accessed from the Download page.\n",
    "\n",
    "## Benchmarking the L1000 Signatures\n",
    "The characteristic direction signatures were benchmarked against three other differential gene-expression analysis methods: fold change, limma [2], and MODZ [3]. Signatures corresponding to the perturbation of cells by the drug dexamethasone were chosen for benchmarking. Specifically, the chosen perturbations are 10 uM treatments of dexamethasone, followed by expression profiling at 24 hours of exposure in the A549 cell line. Each signature was computed by each of the four differential gene expression analysis methods, including the characteristic direction. The top up- and down-regulated genes were identified for each method applied to each signature, and then compared to the NR3C1/Glucocorticoid Receptor target gene set from the ENCODE ChIP-seq library available from ChEA3 [4], as well as to manually computed up and down gene sets from all dexamethasone-related GEO studies from CREEDS [5]. Overlap between the signature-ranked genes and the gene sets was visualized with bridge plots.\n",
    "\n",
    "The bridge plots below show the results after the signatures computed using each of the four methods from all dexamethasone signatures were averaged to generate a single line corresponding to a single differential gene expression characterization method. In general, the characteristic direction method performs better than the MODZ method and is it comparable to limma. It is important to note that the L1000 assay only profiles the expression of 978 genes while the expression of ~12,000 genes is inferred. By comparing these data to independent data that is expected to have some concordance we expect that the methods that increase such concordance are better. However, other benchmarks should be further applied."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![bechmarking.png](img/bechmarking.png)\n",
    "**Figure 2** Benchmarking L1000 Signatures <a name=\"benchmark\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Web Interface\n",
    "### Signature Search\n",
    "![signature_search.png](img/signature-search-workflow.png)\n",
    "**Figure 3** The Signature Search Workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SigCom LINCS's signature search enables users to perform signature similarity searches across over 1.5 million L1000 and other signatures. Users can input up and down genes which can be validated against all the genes that are present in the SigCom LINCS database. This automated validation also provides synonym suggestions for non-standard gene names. Enrichment analysis is performed against seven LINCS datasets, with each dataset corresponding to a specific perturbation type. The top mimicker and reverser signatures are then displayed as bar charts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"img/sigsearch-input-validation.png\" style=\"float: left; width: 50%; margin-right: 25px\"></img><br/><br/><br/><br/><br/>**Figure 4** Gene terms entered as queries to the up and down input boxes are validated automatically against the genes within the SigCom LINCS database. Genes are labeled based on whether they are found in the database, while suggestions are provided for non-standard gene names and synonyms."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"img/signature-search-pertsearch.png\" style=\"float: right; width: 50%; margin-leftt: 25px\"></img><br/><br/><br/><br/><br/>**Figure 5** Users can query SigCom LINCS for gene and chemical perturbagens. The first option, (1) converts a gene into two sets of the mostly positively and negative correlated co-expressed genes based on data complied from the ARCHS4 resource [5]. These top co-expressed genes are used as the up and down input gene sets. The second option, (2) finds signatures where a single gene was silenced using CRISPR. The third option (3) searches for chemically perturbed signatures, and last, (4) returns signatures that maximally up or down regulate the input gene (more information is provided below under the metadata search subheading)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"img/signature-search-results.png\" style=\"float: left; width: 50%; margin-right: 25px\"></img><br/><br/><br/><br/><br/><br/><br/><br/><br/><br/>**Figure 6** The top mimicker and reverser signatures for each query input signature are displayed as bar charts. Hovering over the bars in the chart displays a tooltip with more information about the signature, for example,  the signature full name and the enrichment analysis z-scores."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"img/signature-search-expanded-results.png\" style=\"float: right; width: 50%; margin-left: 25px\"></img><br/><br/><br/><br/><br/><br/><br/><br/><br/><br/><br/><br/>**Figure 7 Expanded view**. Each panel of the results from a query is expandable. Upon expansion, users can view the top mimickers and reversers in one bar chart and two tables. A dedicated search bar is made available for filtering the search results by term. Download buttons are also provided for downloading the tables.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"img/signature-search-scatterplot.png\" style=\"float: left; width: 50%; margin-right: 25px\"></img><br/><br/><br/>**Figure 8 Scatter plots visualization of the enrichment results**. Users can view the signature search results on a scatter plot with the z-up and z-down scores as the x and y axis respectively. On these plots each point represents a matching signature. Signatures located on the top right quadrant are the mimickers while those in the bottom left quadrant are the reversers. Hovering over a point on the plot shows the metadata of the signature as well as the z-scores. Users can change how the points are colored by clicking on the radio buttons at the right. Furthermore, users can perform signature search against signatures with a specific perturbagen either by (1) clicking on the top perturbagen, (2) searching for a perturbagen using the text field, or (3) clicking on a node with the desired perturbagen. This search provides a view about how consistent this perturbagen is in inducing the same changes across many conditions.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Performing Metadata Search\n",
    "SigCom LINCS's metadata search enables users to search the content of the LINCS dataset to identify available datasets to download and analyze. Users can input any search term and then they are presented with matching entries from the SigCom LINCS dataset. Once results are returned, further filtering is provided to narrow down the search. \n",
    "\n",
    "<img src=\"img/metadata-search-results.png\" style=\"float: left; width: 50%; margin-right: 25px\"></img><br/><br/><br/><br/><br/><br/><br/><br/><br/>**Figure 9 Metadata search.** SigCom LINCS provides a metadata search engine that enables users to search for datasets, signatures, or genes by submitting any search term. Such search terms can be for example, an assay, a cell line, a drug, a disease, or a gene name. The search results can be further refined and filtered by  sub setting the results DSGCs, datasets, cell lines, and perturbation. Furthermore, for signature queries with less than 50 results, users can use the provided appyter [6] button to send it for consensus analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"img/metadata-search-options.png\" style=\"float: right; width: 50%; margin-leftt: 25px\"></img><br/><br/>**Figure 10**\n",
    "For L1000 metadata signature search results, users have the option to download the the returned signatures as (1) full rank signature file, or (2) a GMT file containing the top up- and down-regulated genes. Furthermore, users have the option to apply these top up- and down-regulated genes as input for signature search, or send them to Enrichr for enrichment analysis [7-9]."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"img/metadata-search-insignia.png\" style=\"float: left; width: 50%; margin-right: 25px\"></img><br/><br/><br/>**Figure 11** L1000 datasets were assessed for FAIRness [10], these assessments are provided as FAIRshake insignias [11] that can be viewes when searching for datasets."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metadata Dedicated Landing Pages\n",
    "When a user clicks on the results from the metadata search, the entries are brought to the respective metadata page of that entry. Depending on the dataset type, this dedicated page will display the metadata of the entry, as well as the entry's related signatures, genes, and datasets.\n",
    "#### Gene Landing Pages\n",
    "<img src=\"img/metadata-page-gene.png\" style=\"float: left; width: 50%; margin-right: 25px\"></img><br/><br/><br/><br/><br/><br/>**Figure 12** Gene landing pages that contain the signatures that maximally up- or down-regulate the gene are provided by SigCom LINCS. Results from these pages can be filtered by DSGC, dataset, cell line, or perturbagen."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Signature Landing Pages\n",
    "<img src=\"img/metadata-page-signature.png\" style=\"float: right; width: 50%; margin-leftt: 25px\"></img><br/><br/><br/><br/><br/><br/><br/><br/>**Figure 13**\n",
    "Signature dedicated landing pages in SigCom LINCS displays the available metadata for each signature as well as the top up- and down-regulated genes for each signature."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DSGC Dedicated Landing Pages\n",
    "<img src=\"img/metadata-page-dataset.png\" style=\"float: left; width: 50%; margin-right: 25px\"></img><br/><br/><br/><br/><br/><br/>**Figure 14** Dataset pages shows the metadata of the dataset along with the relevant download links, as well as the signatures (if any) that is under that dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DSGC Dedicated Landing Pages\n",
    "<img src=\"img/metadata-page-dsgc.png\" style=\"float: right; width: 50%; margin-leftt: 25px\"></img><br/><br/><br/><br/><br/>**Figure 15** DSGC dedicated landing pages can be accessed from the DSGCs tab. These pages list the available datasets produced by each LINCS DSGC."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The Download Tab\n",
    "The DCIC processed LINCS L1000 data and the other LINCS dataset packages can be  downloaded from the SigCom LINCS download page. Users can download the coefficient tables, GMT files of the up and down gene sets, as well as the predicted RNA-Seq profiles created with applying a deep learning algorithm (CycleGAN) [12].\n",
    "![download.png](img/download.png)\n",
    "**Figure 16** Screenshot from the SigCom LINCS Download Page"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The API Page\n",
    "SigCom LINCS can be accessed via well-documented RESTful APIs. These API are documented with OpenAPI and smartAPI [13]. This documentation can be viewed by going to the API page.Figure 17 Screenshot from the SigCom LINCS API page\n",
    "![API.png](img/api.png)\n",
    "**Figure 17** Screenshot from the SigCom LINCS API page\n",
    "\n",
    "The next section provides examples on how to use the SigCom LINCS RESTful APIs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using the APIs\n",
    "In this section, we describe the two types of RESTful APIs provided by SigCom LINCS for users who wish to access the LINCS data programmatically. The metadata API (https://maayanlab.cloud/sigcom-lincs/metadata-api) provides fast full-text searches as well as querying of data aggregates. Metadata searches are structured using LoopBack 4 queries. The data API (https://maayanlab.cloud/sigcom-lincs/data-api), on the other hand, provides enrichment analysis against LINCS signatures.\n",
    "\n",
    "Here, we explore several use cases and provide Python code that users can use as a template for more complex queries. To start, make sure you have the request library installed via pip. More technical information about the SigCom LINCS API is available from here: https://maayanlab.cloud/sigcom-lincs/#/API.\n",
    "\n",
    "### Full-text metadata search\n",
    "\n",
    "Users can utilize the full-text search capabilities of SigCom LINCS to search for datasets, signatures, or genes using the following URL https://maayanlab.cloud/sigcom-lincs/metadata-api/<model>/find. Below are the mappings available:\n",
    "- DSGCs: resources\n",
    "- Datasets: libraries\n",
    "- Signatures: signatures\n",
    "- Genes: entities\n",
    "\n",
    "Suppose we want to search for datasets (libraries) that contains the word proteomics. We structure our query as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "  {\n",
      "    \"$validator\": \"/dcic/signature-commons-schema/v5/core/library.json\",\n",
      "    \"id\": \"b745f1cf-cd38-4c38-ace3-bdfa1e224586\",\n",
      "    \"resource\": \"09d107aa-4273-43c4-8f03-f985ff9041d1\",\n",
      "    \"dataset\": \"LINCS targeted proteomics signatures\",\n",
      "    \"dataset_type\": \"rank_matrix\",\n",
      "    \"meta\": {\n",
      "      \"icon\": \"static/images/lincs/lincs-pccse.png\",\n",
      "      \"libraryID\": \"LIB_9\",\n",
      "      \"$validator\": \"/@dcic/signature-commons-schema/core/unknown.json\",\n",
      "      \"libraryInfo\": \"Signatures of perturbations assayed by P100 against 96 phosphopeptide probes and GCP assay against ~60 probes that monitor combinations of post-translational modifications on histones. The data is generated by using mass spectrometry techniques to characterize proteome level molecular signatures of responses to small molecule and genetic perturbations in a number of different cell lines.\",\n",
      "      \"libraryName\": \"LINCS targeted proteomics signatures\",\n",
      "      \"sigidPrefix\": \"LINCSTP_\",\n",
      "      \"displayOrder\": 5,\n",
      "      \"signatureCount\": 1178,\n",
      "      \"LINCS_Library_ID\": \"LIB_9\"\n",
      "    }\n",
      "  },\n",
      "  {\n",
      "    \"$validator\": \"/dcic/signature-commons-schema/v5/core/library.json\",\n",
      "    \"id\": \"373ba0e7-787d-4c56-ac06-4a4cc78fb16c\",\n",
      "    \"resource\": \"09c35f9d-76cd-4cec-add5-427605f75254\",\n",
      "    \"dataset\": \"LDS-1546\",\n",
      "    \"dataset_type\": \"rank_matrix\",\n",
      "    \"meta\": {\n",
      "      \"date\": \"2018-10-18\",\n",
      "      \"icon\": \"./static/images/lincs/hms-lincs.png\",\n",
      "      \"size\": \"2.46 MB\",\n",
      "      \"assay\": \"Tandem Mass Tag (TMT) Mass Spectrometry\",\n",
      "      \"center\": \"HMS LINCS (Harvard Medical School)\",\n",
      "      \"$validator\": \"https://raw.githubusercontent.com/MaayanLab/sigcom-lincs/main/validators/lincs_datasets.json\",\n",
      "      \"total_size\": 2464608,\n",
      "      \"datalevel_3\": {\n",
      "        \"id\": \"LDS-1546\",\n",
      "        \"date\": \"2018-10-18\",\n",
      "        \"link\": \"http://lincsportal.ccs.miami.edu/datasets/view/LDS-1546\",\n",
      "        \"size\": \"2.46 MB\",\n",
      "        \"version\": \"1.0\",\n",
      "        \"file_url\": \"https://lincs-dcic.s3.amazonaws.com/LINCS-data-portal/LDS-1546.tar.gz\",\n",
      "        \"filesize\": 2464608,\n",
      "        \"datalevel\": 3,\n",
      "        \"description\": \"Breast Cancer Profiling Project - Proteomics 1: 1 total proteome dataset for a 35-cell line breast cancer panel under basal conditions\"\n",
      "      },\n",
      "      \"description\": \"Breast Cancer Profiling Project - Proteomics 1: 1 total proteome dataset for a 35-cell line breast cancer panel under basal conditions\"\n",
      "    }\n",
      "  }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "\n",
    "API_URL = \"https://maayanlab.cloud/sigcom-lincs/metadata-api/libraries/find\"\n",
    "\n",
    "payload = {\n",
    "    \"filter\": {\n",
    "        \"where\": {\n",
    "            \"meta\": {\n",
    "                \"fullTextSearch\": \"proteomics\"\n",
    "            }\n",
    "        },\n",
    "        \"limit\": 2\n",
    "    }\n",
    "}\n",
    "\n",
    "res = requests.post(API_URL, json=payload)\n",
    "results = res.json()\n",
    "\n",
    "print(json.dumps(results, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtering by metadata fields\n",
    "\n",
    "SigCom LINCS stores its metadata as semi-structured JSON serialized entries. Because of this, we can also filter results via metadata fields. Here we show how to find signatures that are perturbed with a **CRISPR Knockdown**:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "  {\n",
      "    \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "    \"id\": \"d53437f6-2fca-40f8-9a73-5d0b91cd7ef5\",\n",
      "    \"library\": \"b745f1cf-cd38-4c38-ace3-bdfa1e224586\",\n",
      "    \"meta\": {\n",
      "      \"time\": \"23d\",\n",
      "      \"organism\": \"human\",\n",
      "      \"cell_line\": \"YAPC-311\",\n",
      "      \"libraryid\": \"LIB_9\",\n",
      "      \"pert_type\": \"CRISPR Knockdown\",\n",
      "      \"treatment\": \"ASH1L\",\n",
      "      \"$validator\": \"/@dcic/signature-commons-schema/meta/signature/ilincs.json\",\n",
      "      \"signatureid\": \"LINCSTP_10213\",\n",
      "      \"concentration\": \"15152827.96titer\",\n",
      "      \"perturbagenID\": \"BRDN0001057040\",\n",
      "      \"concordancetable\": \"concordanceLib9\"\n",
      "    }\n",
      "  },\n",
      "  {\n",
      "    \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "    \"id\": \"ff68b58c-ed50-4803-ab6d-632cc3eb4f91\",\n",
      "    \"library\": \"b745f1cf-cd38-4c38-ace3-bdfa1e224586\",\n",
      "    \"meta\": {\n",
      "      \"time\": \"12d\",\n",
      "      \"organism\": \"human\",\n",
      "      \"cell_line\": \"A375-311\",\n",
      "      \"libraryid\": \"LIB_9\",\n",
      "      \"pert_type\": \"CRISPR Knockdown\",\n",
      "      \"treatment\": \"CHD6\",\n",
      "      \"$validator\": \"/@dcic/signature-commons-schema/meta/signature/ilincs.json\",\n",
      "      \"signatureid\": \"LINCSTP_10008\",\n",
      "      \"concentration\": \"8205423.278titer\",\n",
      "      \"perturbagenID\": \"BRDN0001057195\",\n",
      "      \"concordancetable\": \"concordanceLib9\"\n",
      "    }\n",
      "  }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "\n",
    "API_URL = \"https://maayanlab.cloud/sigcom-lincs/metadata-api/signatures/find\"\n",
    "\n",
    "payload = {\n",
    "    \"filter\": {\n",
    "        \"where\": {\n",
    "            \"meta.pert_type\":  \"CRISPR Knockdown\"\n",
    "        },\n",
    "        \"limit\": 2\n",
    "    }\n",
    "}\n",
    "\n",
    "res = requests.post(API_URL, json=payload)\n",
    "results = res.json()\n",
    "\n",
    "print(json.dumps(results, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finding available fields\n",
    " Since SigCom LINCS stores metadata as semi-structured JSON serialized entries, this means that although JSON serialization provides opportunities for diverse metadata fields, we still follow certain structures for SigCom LINCS which are defined [here](https://github.com/MaayanLab/sigcom-lincs/tree/main/validators). These structures enable performing queries such as the previous example. In that example, we are sure that the field meta.pert_type exists, and is in fact [required](https://github.com/MaayanLab/sigcom-lincs/blob/a0e6403a668e8397f190fa4e9f398dae7485f0cb/validators/l1000_signatures.json#L94). An easy way to get the available fields for querying without going through the validators is to fetch them using `https://maayanlab.cloud/sigcom-lincs/metadata-api/<model>/key_count`. This returns the available keys, and how many entries have that field. For example, we want to view the available search keys for the genes (entities):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "  \"meta.synonyms\",\n",
      "  \"id\",\n",
      "  \"uuid\",\n",
      "  \"meta.$validator\",\n",
      "  \"meta.geneid\",\n",
      "  \"meta.maplocation\",\n",
      "  \"meta.nomstatus\",\n",
      "  \"meta.taxid\",\n",
      "  \"meta.namenomauth\",\n",
      "  \"meta.dbxrefs\",\n",
      "  \"meta.description\",\n",
      "  \"meta.locustag\",\n",
      "  \"meta.symbol\",\n",
      "  \"meta.chromosome\",\n",
      "  \"meta.symbonomauth\"\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "\n",
    "API_URL = \"https://maayanlab.cloud/sigcom-lincs/metadata-api/entities/key_count\"\n",
    "\n",
    "payload = {\n",
    "    \"limit\": 15\n",
    "}\n",
    "\n",
    "res = requests.get(API_URL, params={\"filter\": json.dumps(payload)})\n",
    "results = res.json()\n",
    "\n",
    "print(json.dumps(list(results.keys()), indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Counting the results <a name=\"count\"></a>\n",
    "\n",
    "The count endpoint `https://maayanlab.cloud/sigcom-lincs/metadata-api/<model>/count` accepts GET requests to count the number of entries in a model. Users can also pass a `where` filter to get a filtered count. Here we show how to get the number of datasets (libraries) that use the ***L1000 mRNA profiling assay***."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"count\": 19\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "\n",
    "API_URL = \"https://maayanlab.cloud/sigcom-lincs/metadata-api/libraries/count\"\n",
    "\n",
    "payload = {\n",
    "    \"meta.assay\": \"L1000 mRNA profiling assay\"\n",
    "}\n",
    "\n",
    "res = requests.get(API_URL, params={\"where\": json.dumps(payload)})\n",
    "results = res.json()\n",
    "\n",
    "print(json.dumps(results, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fetching the value counts\n",
    "The value count endpoint `https://maayanlab.cloud/sigcom-lincs/metadata-api/<model>/value_count` is used to count the values of a specific field. This is particularly useful for obtaining the top assays, cell line, perturbations of a model. Below we show how to get the top 25 cell lines of the signatures perturbed with ***dexamethasone***."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"meta.cell_line\": {\n",
      "    \"MCF7\": 46,\n",
      "    \"A549\": 39,\n",
      "    \"PC3\": 34,\n",
      "    \"A375\": 33,\n",
      "    \"HA1E\": 27,\n",
      "    \"HT29\": 23,\n",
      "    \"YAPC\": 16,\n",
      "    \"HTB-22\": 14,\n",
      "    \"VCAP\": 14,\n",
      "    \"CRL-1619\": 13,\n",
      "    \"CRL-1435\": 13,\n",
      "    \"JURKAT\": 12,\n",
      "    \"MDAMB231\": 12,\n",
      "    \"MCF10A\": 12,\n",
      "    \"HPTEC\": 12,\n",
      "    \"HEK293\": 12,\n",
      "    \"HELA\": 12,\n",
      "    \"HCC515\": 11,\n",
      "    \"U2OS\": 11,\n",
      "    \"HTB-38\": 10,\n",
      "    \"NPC\": 10,\n",
      "    \"THP1\": 8,\n",
      "    \"HEPG2\": 8,\n",
      "    \"CCL-185\": 7,\n",
      "    \"CCL-2\": 6\n",
      "  }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "\n",
    "API_URL = \"https://maayanlab.cloud/sigcom-lincs/metadata-api/signatures/value_count\"\n",
    "\n",
    "payload = {\n",
    "    \"where\": {\n",
    "        \"meta.pert_name\":  \"dexamethasone\"\n",
    "    },\n",
    "    \"fields\": [\"meta.cell_line\"],\n",
    "    \"limit\": 25\n",
    "}\n",
    "\n",
    "res = requests.get(API_URL, params={\"filter\": json.dumps(payload)})\n",
    "results = res.json()\n",
    "\n",
    "print(json.dumps(results, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Performing Signature Search\n",
    "Because the data API is separate from the metadata API, signature search is a multi-step process:\n",
    "\n",
    "#### Step 1: Convert gene names to UUIDs using the metadata API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"up_entities\": [\n",
      "    \"41f5b538-2b08-4286-86f3-3e88470ed141\",\n",
      "    \"8088700c-6942-4c4b-8ee2-3dc67b7a1a45\",\n",
      "    \"3d768c6e-b3fa-4dd2-b740-540e03e29384\",\n",
      "    \"e9e7aaca-1b14-4f1b-af61-6deb85d90399\",\n",
      "    \"c52c0f4f-4696-4d91-8c15-c9245546787a\",\n",
      "    \"e91105d5-a9a7-4a39-bff4-a8cbe685bc08\",\n",
      "    \"4c4bfc3f-5455-4327-bfbe-3dcd3cc8091b\"\n",
      "  ],\n",
      "  \"down_entities\": [\n",
      "    \"566ef317-6636-4993-8a5b-341b00c70498\",\n",
      "    \"d23d4236-6a13-4bda-9c53-f5b6e0f13263\",\n",
      "    \"ca7488f8-3265-4872-8bea-67a571cfa23d\",\n",
      "    \"a743b103-c306-48d5-a3cf-36a2fedd379e\",\n",
      "    \"5fc61a62-f46d-4a15-953e-90daca7e73d2\",\n",
      "    \"c70cac7b-35df-41b5-9295-fa6141071c7a\",\n",
      "    \"16c838f0-f8f7-401f-919a-30b139a7d7cb\"\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "\n",
    "METADATA_API = \"https://maayanlab.cloud/sigcom-lincs/metadata-api/\"\n",
    "DATA_API = \"https://maayanlab.cloud/sigcom-lincs/data-api/api/v1/\"\n",
    "\n",
    "input_gene_set = {\n",
    "    \"up_genes\": [\"TARBP1\", \"APP\", \"RAP1GAP\", \"UFM1\", \"DNAJA3\", \"PCBD1\", \"CSRP1\"],\n",
    "    \"down_genes\": [\"CEBPA\", \"STAT5B\", \"DSE\", \"EIF4EBP1\", \"CARD8\", \"HLA-DMA\", \"SERPINE1\"]\n",
    "}\n",
    "\n",
    "all_genes = input_gene_set[\"up_genes\"] + input_gene_set[\"down_genes\"]\n",
    "\n",
    "payload = {\n",
    "    \"filter\": {\n",
    "        \"where\": {\n",
    "            \"meta.symbol\": {\n",
    "                \"inq\": all_genes\n",
    "            }\n",
    "        },\n",
    "        \"fields\": [\"id\", \"meta.symbol\"]\n",
    "    }\n",
    "}\n",
    "res = requests.post(METADATA_API + \"entities/find\", json=payload)\n",
    "entities = res.json()\n",
    "\n",
    "for_enrichment = {\n",
    "    \"up_entities\": [],\n",
    "    \"down_entities\": []\n",
    "}\n",
    "\n",
    "for e in entities:\n",
    "    symbol = e[\"meta\"][\"symbol\"]\n",
    "    if symbol in input_gene_set[\"up_genes\"]:\n",
    "        for_enrichment[\"up_entities\"].append(e[\"id\"])\n",
    "    elif symbol in input_gene_set[\"down_genes\"]:\n",
    "        for_enrichment[\"down_entities\"].append(e[\"id\"])\n",
    "\n",
    "print(json.dumps(for_enrichment, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 2 Perform signature search using the ranktwosided endpoint of the data API\n",
    "For general use, the ranktwosided endpoint takes as input the up and down gene sets, a limit, and the database you want to use for enrichment. Obtaining the list of available databases for enrichment is discussed in the next section. The data API returns the top matching signatures along with the enrichment scores, ranked by the sum of the z-up and z-down scores (z-sum). A positive z-sum indicates that the signature is a mimicker while a negative z-sum means it is a reverser."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"queryTimeSec\": 0.327,\n",
      "  \"results\": [\n",
      "    {\n",
      "      \"direction-up\": -1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 6.645764878704664,\n",
      "      \"logp-fisher\": 13.298411272025652,\n",
      "      \"p-up\": 0.0011918982073126916,\n",
      "      \"uuid\": \"cb03f2fd-2ce7-5ee7-85f1-dff255d942b0\",\n",
      "      \"z-down\": 3.1932109963550976,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": -3.2408415054818382,\n",
      "      \"p-down\": 0.0014071299274773796,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": 1\n",
      "    },\n",
      "    {\n",
      "      \"direction-up\": 1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 5.454303139526856,\n",
      "      \"logp-fisher\": 13.42078558581136,\n",
      "      \"p-up\": 0.008378598563641315,\n",
      "      \"uuid\": \"a1d5e0de-4a40-5523-998b-145a1406ff3f\",\n",
      "      \"z-down\": -3.7496799355724173,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": 2.6364231028355904,\n",
      "      \"p-down\": 0.00017711506152195966,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": -1\n",
      "    },\n",
      "    {\n",
      "      \"direction-up\": -1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 6.376111633289184,\n",
      "      \"logp-fisher\": 12.76538406482507,\n",
      "      \"p-up\": 0.0018963094983097672,\n",
      "      \"uuid\": \"b60f0d4c-6b6c-5062-b5bf-b7abfe891d58\",\n",
      "      \"z-down\": -3.173329511518713,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": -3.1060301537570463,\n",
      "      \"p-down\": 0.0015071445967729247,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": -1\n",
      "    },\n",
      "    {\n",
      "      \"direction-up\": -1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 5.753724899245878,\n",
      "      \"logp-fisher\": 12.77087535497806,\n",
      "      \"p-up\": 0.00048532905810971805,\n",
      "      \"uuid\": \"cfba7dd3-317c-54d5-905f-cba9b9f06b62\",\n",
      "      \"z-down\": 2.7557120118539,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": -3.4887753163826387,\n",
      "      \"p-down\": 0.005856565565235616,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": 1\n",
      "    },\n",
      "    {\n",
      "      \"direction-up\": -1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 6.009853708923177,\n",
      "      \"logp-fisher\": 12.280623714067888,\n",
      "      \"p-up\": 0.003630674568347647,\n",
      "      \"uuid\": \"e51488e8-472d-5f7d-a7de-66486f1da161\",\n",
      "      \"z-down\": 3.220853702544724,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": -2.9085974407026782,\n",
      "      \"p-down\": 0.0012782198674079481,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": 1\n",
      "    },\n",
      "    {\n",
      "      \"direction-up\": 1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 6.0542185234053125,\n",
      "      \"logp-fisher\": 12.165580256107136,\n",
      "      \"p-up\": 0.0029012786175475203,\n",
      "      \"uuid\": \"28cb409c-fd34-5980-ad3c-3ab5dfa56fbb\",\n",
      "      \"z-down\": 3.122296823168634,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": 2.9780231604789313,\n",
      "      \"p-down\": 0.001794593904739905,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": 1\n",
      "    },\n",
      "    {\n",
      "      \"direction-up\": 1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 5.103069712945671,\n",
      "      \"logp-fisher\": 12.448437603385242,\n",
      "      \"p-up\": 0.0003318469186532891,\n",
      "      \"uuid\": \"4eb1a739-5090-55c8-b484-49db420eb00c\",\n",
      "      \"z-down\": 2.5173468300187394,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": 3.589139603471127,\n",
      "      \"p-down\": 0.011824273080401904,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": 1\n",
      "    },\n",
      "    {\n",
      "      \"direction-up\": -1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 5.85585523897824,\n",
      "      \"logp-fisher\": 11.803099321863376,\n",
      "      \"p-up\": 0.0020169617276937224,\n",
      "      \"uuid\": \"dedf3bef-c834-57b7-a9cf-99a81cf826a0\",\n",
      "      \"z-down\": 2.90189940035673,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": -3.087743440431601,\n",
      "      \"p-down\": 0.0037092101344076145,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": 1\n",
      "    },\n",
      "    {\n",
      "      \"direction-up\": 1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 5.764767008424871,\n",
      "      \"logp-fisher\": 11.723466309786144,\n",
      "      \"p-up\": 0.004452874409889285,\n",
      "      \"uuid\": \"0d7566c1-d8b7-53e6-b987-b9dc58d2ea54\",\n",
      "      \"z-down\": -3.1182567353409194,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": 2.8441686716607033,\n",
      "      \"p-down\": 0.0018193775293000591,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": -1\n",
      "    },\n",
      "    {\n",
      "      \"direction-up\": 1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 5.1935033396042405,\n",
      "      \"logp-fisher\": 12.101688714246226,\n",
      "      \"p-up\": 0.0005245635594222087,\n",
      "      \"uuid\": \"d572b395-a6e0-5c6d-9f60-80635e6a25d6\",\n",
      "      \"z-down\": 2.5562592548856746,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": 3.4679369686396897,\n",
      "      \"p-down\": 0.010580477113730336,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": 1\n",
      "    }\n",
      "  ],\n",
      "  \"signatures\": []\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "query = {\n",
    "    **for_enrichment,\n",
    "    \"limit\": 10,\n",
    "    \"database\": \"l1000_xpr\"\n",
    "}\n",
    "\n",
    "res = requests.post(DATA_API + \"enrich/ranktwosided\", json=query)\n",
    "results = res.json()\n",
    "\n",
    "# Optional, multiply z-down and direction-down with -1\n",
    "for i in results[\"results\"]:\n",
    "    i[\"z-down\"] = -i[\"z-down\"]\n",
    "    i[\"direction-down\"] = -i[\"direction-down\"]\n",
    "print(json.dumps(results, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 3: Resolve signature UUIDs using the metadata API\n",
    "We can now resolve the UUIDs of the returned signatures using the metadata API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "  {\n",
      "    \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "    \"id\": \"0d7566c1-d8b7-53e6-b987-b9dc58d2ea54\",\n",
      "    \"library\": \"96c7b8c5-1eca-5764-88e4-e4ccaee6603f\",\n",
      "    \"meta\": {\n",
      "      \"md5\": \"412d5750f07e026a92db9c3ebc4ddd16\",\n",
      "      \"sha256\": \"6c7b8ec5000429ce3b2572440a7343b928142ca1ac40edb47f1e6c2fc2a3ad3d\",\n",
      "      \"tissue\": \"skin of body\",\n",
      "      \"anatomy\": \"UBERON:0002097\",\n",
      "      \"cmap_id\": \"XPR045_HS944T.311_96H:M15\",\n",
      "      \"version\": 1,\n",
      "      \"filename\": \"L1000_LINCS_DCIC_XPR045_HS944T.311_96H_M15_BRDN0001145659.tsv\",\n",
      "      \"local_id\": \"XPR045_HS944T.311_96H_M15_BRDN0001145659\",\n",
      "      \"cell_line\": \"HS944T\",\n",
      "      \"pert_name\": \"BRDN0001145659\",\n",
      "      \"pert_time\": \"96 h\",\n",
      "      \"pert_type\": \"CRISPR Knockdown\",\n",
      "      \"$validator\": \"https://raw.githubusercontent.com/MaayanLab/sigcom-lincs/main/validators/l1000_signatures.json\",\n",
      "      \"data_level\": 5,\n",
      "      \"creation_time\": \"2021-05-26\",\n",
      "      \"persistent_id\": \"https://lincs-dcic.s3.amazonaws.com/LINCS-sigs-2021/cd/xpr/L1000_LINCS_DCIC_XPR045_HS944T.311_96H_M15_BRDN0001145659.tsv\",\n",
      "      \"size_in_bytes\": 216853,\n",
      "      \"uncompressed_size_in_bytes\": 216853\n",
      "    },\n",
      "    \"scores\": {\n",
      "      \"direction-up\": 1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 5.764767008424871,\n",
      "      \"logp-fisher\": 11.723466309786144,\n",
      "      \"p-up\": 0.004452874409889285,\n",
      "      \"z-down\": -3.1182567353409194,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": 2.8441686716607033,\n",
      "      \"p-down\": 0.0018193775293000591,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": -1\n",
      "    }\n",
      "  },\n",
      "  {\n",
      "    \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "    \"id\": \"28cb409c-fd34-5980-ad3c-3ab5dfa56fbb\",\n",
      "    \"library\": \"96c7b8c5-1eca-5764-88e4-e4ccaee6603f\",\n",
      "    \"meta\": {\n",
      "      \"md5\": \"5c9bea5c57b3ee955e6f533f93afd4db\",\n",
      "      \"sha256\": \"a588b00dc1642960a2fa3fb6fd128564d24b532a547d8bd5ae1509d5040e1c0a\",\n",
      "      \"tissue\": \"stomach\",\n",
      "      \"anatomy\": \"UBERON:0000945\",\n",
      "      \"cmap_id\": \"XPR015_AGS.311_96H:P08\",\n",
      "      \"version\": 1,\n",
      "      \"filename\": \"L1000_LINCS_DCIC_XPR015_AGS.311_96H_P08_DNAJC16.tsv\",\n",
      "      \"local_id\": \"XPR015_AGS.311_96H_P08_DNAJC16\",\n",
      "      \"cell_line\": \"AGS\",\n",
      "      \"pert_name\": \"DNAJC16\",\n",
      "      \"pert_time\": \"96 h\",\n",
      "      \"pert_type\": \"CRISPR Knockdown\",\n",
      "      \"$validator\": \"https://raw.githubusercontent.com/MaayanLab/sigcom-lincs/main/validators/l1000_signatures.json\",\n",
      "      \"data_level\": 5,\n",
      "      \"creation_time\": \"2021-05-24\",\n",
      "      \"persistent_id\": \"https://lincs-dcic.s3.amazonaws.com/LINCS-sigs-2021/cd/xpr/L1000_LINCS_DCIC_XPR015_AGS.311_96H_P08_DNAJC16.tsv\",\n",
      "      \"size_in_bytes\": 217232,\n",
      "      \"uncompressed_size_in_bytes\": 217232\n",
      "    },\n",
      "    \"scores\": {\n",
      "      \"direction-up\": 1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 6.0542185234053125,\n",
      "      \"logp-fisher\": 12.165580256107136,\n",
      "      \"p-up\": 0.0029012786175475203,\n",
      "      \"z-down\": 3.122296823168634,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": 2.9780231604789313,\n",
      "      \"p-down\": 0.001794593904739905,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": 1\n",
      "    }\n",
      "  },\n",
      "  {\n",
      "    \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "    \"id\": \"4eb1a739-5090-55c8-b484-49db420eb00c\",\n",
      "    \"library\": \"96c7b8c5-1eca-5764-88e4-e4ccaee6603f\",\n",
      "    \"meta\": {\n",
      "      \"md5\": \"b2236b54f3f24e0e377a19b269586487\",\n",
      "      \"sha256\": \"6b14b231780fff988a7eac102f950fe4c31d0dcc7c4f3eef034f0793a6d4d1c6\",\n",
      "      \"tissue\": \"skin of body\",\n",
      "      \"anatomy\": \"UBERON:0002097\",\n",
      "      \"cmap_id\": \"XPR044_IPC298.311_96H:F19\",\n",
      "      \"version\": 1,\n",
      "      \"filename\": \"L1000_LINCS_DCIC_XPR044_IPC298.311_96H_F19_NDUFB7.tsv\",\n",
      "      \"local_id\": \"XPR044_IPC298.311_96H_F19_NDUFB7\",\n",
      "      \"cell_line\": \"IPC298\",\n",
      "      \"pert_name\": \"NDUFB7\",\n",
      "      \"pert_time\": \"96 h\",\n",
      "      \"pert_type\": \"CRISPR Knockdown\",\n",
      "      \"$validator\": \"https://raw.githubusercontent.com/MaayanLab/sigcom-lincs/main/validators/l1000_signatures.json\",\n",
      "      \"data_level\": 5,\n",
      "      \"creation_time\": \"2021-05-27\",\n",
      "      \"persistent_id\": \"https://lincs-dcic.s3.amazonaws.com/LINCS-sigs-2021/cd/xpr/L1000_LINCS_DCIC_XPR044_IPC298.311_96H_F19_NDUFB7.tsv\",\n",
      "      \"size_in_bytes\": 216951,\n",
      "      \"uncompressed_size_in_bytes\": 216951\n",
      "    },\n",
      "    \"scores\": {\n",
      "      \"direction-up\": 1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 5.103069712945671,\n",
      "      \"logp-fisher\": 12.448437603385242,\n",
      "      \"p-up\": 0.0003318469186532891,\n",
      "      \"z-down\": 2.5173468300187394,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": 3.589139603471127,\n",
      "      \"p-down\": 0.011824273080401904,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": 1\n",
      "    }\n",
      "  },\n",
      "  {\n",
      "    \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "    \"id\": \"a1d5e0de-4a40-5523-998b-145a1406ff3f\",\n",
      "    \"library\": \"96c7b8c5-1eca-5764-88e4-e4ccaee6603f\",\n",
      "    \"meta\": {\n",
      "      \"md5\": \"07f5a8648ad832044b2f691f1cfe218c\",\n",
      "      \"sha256\": \"7bfa6568f5306857a1cf3fd9ea4d15fabed55d31b5936bfc9c77e5512357b927\",\n",
      "      \"tissue\": \"hypopharynx\",\n",
      "      \"anatomy\": \"UBERON:0001051\",\n",
      "      \"cmap_id\": \"XPR021_BICR6.311_96H:F13\",\n",
      "      \"version\": 1,\n",
      "      \"filename\": \"L1000_LINCS_DCIC_XPR021_BICR6.311_96H_F13_NHLH2.tsv\",\n",
      "      \"local_id\": \"XPR021_BICR6.311_96H_F13_NHLH2\",\n",
      "      \"cell_line\": \"BICR6\",\n",
      "      \"pert_name\": \"NHLH2\",\n",
      "      \"pert_time\": \"96 h\",\n",
      "      \"pert_type\": \"CRISPR Knockdown\",\n",
      "      \"$validator\": \"https://raw.githubusercontent.com/MaayanLab/sigcom-lincs/main/validators/l1000_signatures.json\",\n",
      "      \"data_level\": 5,\n",
      "      \"creation_time\": \"2021-05-24\",\n",
      "      \"persistent_id\": \"https://lincs-dcic.s3.amazonaws.com/LINCS-sigs-2021/cd/xpr/L1000_LINCS_DCIC_XPR021_BICR6.311_96H_F13_NHLH2.tsv\",\n",
      "      \"size_in_bytes\": 217844,\n",
      "      \"uncompressed_size_in_bytes\": 217844\n",
      "    },\n",
      "    \"scores\": {\n",
      "      \"direction-up\": 1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 5.454303139526856,\n",
      "      \"logp-fisher\": 13.42078558581136,\n",
      "      \"p-up\": 0.008378598563641315,\n",
      "      \"z-down\": -3.7496799355724173,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": 2.6364231028355904,\n",
      "      \"p-down\": 0.00017711506152195966,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": -1\n",
      "    }\n",
      "  },\n",
      "  {\n",
      "    \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "    \"id\": \"b60f0d4c-6b6c-5062-b5bf-b7abfe891d58\",\n",
      "    \"library\": \"96c7b8c5-1eca-5764-88e4-e4ccaee6603f\",\n",
      "    \"meta\": {\n",
      "      \"md5\": \"5af05d3431d77ca3f84c23aba7c99501\",\n",
      "      \"sha256\": \"0f7349bb5f30b026df5ed548f1bc6ba7e8137496b554ee046101f7bc83c8a2f8\",\n",
      "      \"tissue\": \"stomach\",\n",
      "      \"anatomy\": \"UBERON:0000945\",\n",
      "      \"cmap_id\": \"XPR015_AGS.311_96H:L21\",\n",
      "      \"version\": 1,\n",
      "      \"filename\": \"L1000_LINCS_DCIC_XPR015_AGS.311_96H_L21_DACH1.tsv\",\n",
      "      \"local_id\": \"XPR015_AGS.311_96H_L21_DACH1\",\n",
      "      \"cell_line\": \"AGS\",\n",
      "      \"pert_name\": \"DACH1\",\n",
      "      \"pert_time\": \"96 h\",\n",
      "      \"pert_type\": \"CRISPR Knockdown\",\n",
      "      \"$validator\": \"https://raw.githubusercontent.com/MaayanLab/sigcom-lincs/main/validators/l1000_signatures.json\",\n",
      "      \"data_level\": 5,\n",
      "      \"creation_time\": \"2021-05-24\",\n",
      "      \"persistent_id\": \"https://lincs-dcic.s3.amazonaws.com/LINCS-sigs-2021/cd/xpr/L1000_LINCS_DCIC_XPR015_AGS.311_96H_L21_DACH1.tsv\",\n",
      "      \"size_in_bytes\": 217522,\n",
      "      \"uncompressed_size_in_bytes\": 217522\n",
      "    },\n",
      "    \"scores\": {\n",
      "      \"direction-up\": -1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 6.376111633289184,\n",
      "      \"logp-fisher\": 12.76538406482507,\n",
      "      \"p-up\": 0.0018963094983097672,\n",
      "      \"z-down\": -3.173329511518713,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": -3.1060301537570463,\n",
      "      \"p-down\": 0.0015071445967729247,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": -1\n",
      "    }\n",
      "  },\n",
      "  {\n",
      "    \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "    \"id\": \"cb03f2fd-2ce7-5ee7-85f1-dff255d942b0\",\n",
      "    \"library\": \"96c7b8c5-1eca-5764-88e4-e4ccaee6603f\",\n",
      "    \"meta\": {\n",
      "      \"md5\": \"11d222b7587b465b247ac276ded48d98\",\n",
      "      \"sha256\": \"16c71b3e03cc89b3f213383736ee7eb12bbf6ee20e2b48af4a1e2890bbe13c0c\",\n",
      "      \"tissue\": \"hypopharynx\",\n",
      "      \"anatomy\": \"UBERON:0001051\",\n",
      "      \"cmap_id\": \"XPR020_BICR6.311_96H:B09\",\n",
      "      \"version\": 1,\n",
      "      \"filename\": \"L1000_LINCS_DCIC_XPR020_BICR6.311_96H_B09_MED23.tsv\",\n",
      "      \"local_id\": \"XPR020_BICR6.311_96H_B09_MED23\",\n",
      "      \"cell_line\": \"BICR6\",\n",
      "      \"pert_name\": \"MED23\",\n",
      "      \"pert_time\": \"96 h\",\n",
      "      \"pert_type\": \"CRISPR Knockdown\",\n",
      "      \"$validator\": \"https://raw.githubusercontent.com/MaayanLab/sigcom-lincs/main/validators/l1000_signatures.json\",\n",
      "      \"data_level\": 5,\n",
      "      \"creation_time\": \"2021-05-24\",\n",
      "      \"persistent_id\": \"https://lincs-dcic.s3.amazonaws.com/LINCS-sigs-2021/cd/xpr/L1000_LINCS_DCIC_XPR020_BICR6.311_96H_B09_MED23.tsv\",\n",
      "      \"size_in_bytes\": 217021,\n",
      "      \"uncompressed_size_in_bytes\": 217021\n",
      "    },\n",
      "    \"scores\": {\n",
      "      \"direction-up\": -1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 6.645764878704664,\n",
      "      \"logp-fisher\": 13.298411272025652,\n",
      "      \"p-up\": 0.0011918982073126916,\n",
      "      \"z-down\": 3.1932109963550976,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": -3.2408415054818382,\n",
      "      \"p-down\": 0.0014071299274773796,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": 1\n",
      "    }\n",
      "  },\n",
      "  {\n",
      "    \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "    \"id\": \"cfba7dd3-317c-54d5-905f-cba9b9f06b62\",\n",
      "    \"library\": \"96c7b8c5-1eca-5764-88e4-e4ccaee6603f\",\n",
      "    \"meta\": {\n",
      "      \"md5\": \"7661e2ff4f06306236b2be1ed25d66ce\",\n",
      "      \"sha256\": \"24c23baae6f2c0ab508f1c2e04ffc190772b207484172bfb1cddc84e05eebea0\",\n",
      "      \"tissue\": \"hypopharynx\",\n",
      "      \"anatomy\": \"UBERON:0001051\",\n",
      "      \"cmap_id\": \"XPR012_BICR6.311_96H:I07\",\n",
      "      \"version\": 1,\n",
      "      \"filename\": \"L1000_LINCS_DCIC_XPR012_BICR6.311_96H_I07_ASCL5.tsv\",\n",
      "      \"local_id\": \"XPR012_BICR6.311_96H_I07_ASCL5\",\n",
      "      \"cell_line\": \"BICR6\",\n",
      "      \"pert_name\": \"ASCL5\",\n",
      "      \"pert_time\": \"96 h\",\n",
      "      \"pert_type\": \"CRISPR Knockdown\",\n",
      "      \"$validator\": \"https://raw.githubusercontent.com/MaayanLab/sigcom-lincs/main/validators/l1000_signatures.json\",\n",
      "      \"data_level\": 5,\n",
      "      \"creation_time\": \"2021-05-24\",\n",
      "      \"persistent_id\": \"https://lincs-dcic.s3.amazonaws.com/LINCS-sigs-2021/cd/xpr/L1000_LINCS_DCIC_XPR012_BICR6.311_96H_I07_ASCL5.tsv\",\n",
      "      \"size_in_bytes\": 216921,\n",
      "      \"uncompressed_size_in_bytes\": 216921\n",
      "    },\n",
      "    \"scores\": {\n",
      "      \"direction-up\": -1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 5.753724899245878,\n",
      "      \"logp-fisher\": 12.77087535497806,\n",
      "      \"p-up\": 0.00048532905810971805,\n",
      "      \"z-down\": 2.7557120118539,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": -3.4887753163826387,\n",
      "      \"p-down\": 0.005856565565235616,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": 1\n",
      "    }\n",
      "  },\n",
      "  {\n",
      "    \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "    \"id\": \"d572b395-a6e0-5c6d-9f60-80635e6a25d6\",\n",
      "    \"library\": \"96c7b8c5-1eca-5764-88e4-e4ccaee6603f\",\n",
      "    \"meta\": {\n",
      "      \"md5\": \"95553528d976f34744dd4e0c11cf2399\",\n",
      "      \"sha256\": \"4b261b1ea614e59288aafdf0b23b101fbbf69fa8d3fe29ee1fe02741dcb195ba\",\n",
      "      \"tissue\": \"endometrium\",\n",
      "      \"anatomy\": \"UBERON:0001295\",\n",
      "      \"cmap_id\": \"XPR039_SNGM.311_96H:H05\",\n",
      "      \"version\": 1,\n",
      "      \"filename\": \"L1000_LINCS_DCIC_XPR039_SNGM.311_96H_H05_BRDN0001489766.tsv\",\n",
      "      \"local_id\": \"XPR039_SNGM.311_96H_H05_BRDN0001489766\",\n",
      "      \"cell_line\": \"SNGM\",\n",
      "      \"pert_name\": \"BRDN0001489766\",\n",
      "      \"pert_time\": \"96 h\",\n",
      "      \"pert_type\": \"CRISPR Knockdown\",\n",
      "      \"$validator\": \"https://raw.githubusercontent.com/MaayanLab/sigcom-lincs/main/validators/l1000_signatures.json\",\n",
      "      \"data_level\": 5,\n",
      "      \"creation_time\": \"2021-05-26\",\n",
      "      \"persistent_id\": \"https://lincs-dcic.s3.amazonaws.com/LINCS-sigs-2021/cd/xpr/L1000_LINCS_DCIC_XPR039_SNGM.311_96H_H05_BRDN0001489766.tsv\",\n",
      "      \"size_in_bytes\": 217315,\n",
      "      \"uncompressed_size_in_bytes\": 217315\n",
      "    },\n",
      "    \"scores\": {\n",
      "      \"direction-up\": 1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 5.1935033396042405,\n",
      "      \"logp-fisher\": 12.101688714246226,\n",
      "      \"p-up\": 0.0005245635594222087,\n",
      "      \"z-down\": 2.5562592548856746,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": 3.4679369686396897,\n",
      "      \"p-down\": 0.010580477113730336,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": 1\n",
      "    }\n",
      "  },\n",
      "  {\n",
      "    \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "    \"id\": \"dedf3bef-c834-57b7-a9cf-99a81cf826a0\",\n",
      "    \"library\": \"96c7b8c5-1eca-5764-88e4-e4ccaee6603f\",\n",
      "    \"meta\": {\n",
      "      \"md5\": \"d6b8151ff7940006abea7deb3c972646\",\n",
      "      \"sha256\": \"5d4c0bdeb6d527f9630f67a42299bfb0f33c889935d626dd462cea6d91532bee\",\n",
      "      \"tissue\": \"hypopharynx\",\n",
      "      \"anatomy\": \"UBERON:0001051\",\n",
      "      \"cmap_id\": \"XPR023_BICR6.311_96H:L14\",\n",
      "      \"version\": 1,\n",
      "      \"filename\": \"L1000_LINCS_DCIC_XPR023_BICR6.311_96H_L14_SF3A2.tsv\",\n",
      "      \"local_id\": \"XPR023_BICR6.311_96H_L14_SF3A2\",\n",
      "      \"cell_line\": \"BICR6\",\n",
      "      \"pert_name\": \"SF3A2\",\n",
      "      \"pert_time\": \"96 h\",\n",
      "      \"pert_type\": \"CRISPR Knockdown\",\n",
      "      \"$validator\": \"https://raw.githubusercontent.com/MaayanLab/sigcom-lincs/main/validators/l1000_signatures.json\",\n",
      "      \"data_level\": 5,\n",
      "      \"creation_time\": \"2021-05-24\",\n",
      "      \"persistent_id\": \"https://lincs-dcic.s3.amazonaws.com/LINCS-sigs-2021/cd/xpr/L1000_LINCS_DCIC_XPR023_BICR6.311_96H_L14_SF3A2.tsv\",\n",
      "      \"size_in_bytes\": 216863,\n",
      "      \"uncompressed_size_in_bytes\": 216863\n",
      "    },\n",
      "    \"scores\": {\n",
      "      \"direction-up\": -1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 5.85585523897824,\n",
      "      \"logp-fisher\": 11.803099321863376,\n",
      "      \"p-up\": 0.0020169617276937224,\n",
      "      \"z-down\": 2.90189940035673,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": -3.087743440431601,\n",
      "      \"p-down\": 0.0037092101344076145,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": 1\n",
      "    }\n",
      "  },\n",
      "  {\n",
      "    \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "    \"id\": \"e51488e8-472d-5f7d-a7de-66486f1da161\",\n",
      "    \"library\": \"96c7b8c5-1eca-5764-88e4-e4ccaee6603f\",\n",
      "    \"meta\": {\n",
      "      \"md5\": \"93cb7ddc493a01be56c4a6a080d38148\",\n",
      "      \"sha256\": \"98387015b7d10249476c48fc134655a349146eefb4e9b046dd5348588e41d786\",\n",
      "      \"tissue\": \"hypopharynx\",\n",
      "      \"anatomy\": \"UBERON:0001051\",\n",
      "      \"cmap_id\": \"XPR019_BICR6.311_96H:D03\",\n",
      "      \"version\": 1,\n",
      "      \"filename\": \"L1000_LINCS_DCIC_XPR019_BICR6.311_96H_D03_LCP2.tsv\",\n",
      "      \"local_id\": \"XPR019_BICR6.311_96H_D03_LCP2\",\n",
      "      \"cell_line\": \"BICR6\",\n",
      "      \"pert_name\": \"LCP2\",\n",
      "      \"pert_time\": \"96 h\",\n",
      "      \"pert_type\": \"CRISPR Knockdown\",\n",
      "      \"$validator\": \"https://raw.githubusercontent.com/MaayanLab/sigcom-lincs/main/validators/l1000_signatures.json\",\n",
      "      \"data_level\": 5,\n",
      "      \"creation_time\": \"2021-05-24\",\n",
      "      \"persistent_id\": \"https://lincs-dcic.s3.amazonaws.com/LINCS-sigs-2021/cd/xpr/L1000_LINCS_DCIC_XPR019_BICR6.311_96H_D03_LCP2.tsv\",\n",
      "      \"size_in_bytes\": 217047,\n",
      "      \"uncompressed_size_in_bytes\": 217047\n",
      "    },\n",
      "    \"scores\": {\n",
      "      \"direction-up\": -1,\n",
      "      \"fdr-down\": 0.08597436403632452,\n",
      "      \"logp-avg\": 6.009853708923177,\n",
      "      \"logp-fisher\": 12.280623714067888,\n",
      "      \"p-up\": 0.003630674568347647,\n",
      "      \"z-down\": 3.220853702544724,\n",
      "      \"fdr-up\": 0.08631352888700067,\n",
      "      \"z-up\": -2.9085974407026782,\n",
      "      \"p-down\": 0.0012782198674079481,\n",
      "      \"p-down-bonferroni\": 1,\n",
      "      \"p-up-bonferroni\": 1,\n",
      "      \"direction-down\": 1\n",
      "    }\n",
      "  }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "sigids = {i[\"uuid\"]: i for i in results[\"results\"]}\n",
    "\n",
    "payload = {\n",
    "    \"filter\": {\n",
    "        \"where\": {\n",
    "            \"id\": {\n",
    "                \"inq\": list(sigids.keys())\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "res = requests.post(METADATA_API + \"signatures/find\", json=payload)\n",
    "signatures = res.json()\n",
    "\n",
    "## Merge the scores and the metadata\n",
    "for sig in signatures:\n",
    "    uid = sig[\"id\"]\n",
    "    scores = sigids[uid]\n",
    "    scores.pop(\"uuid\")\n",
    "    sig[\"scores\"] = scores\n",
    "\n",
    "print(json.dumps(signatures, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Listing available datasets and databases for search\n",
    "To view the available databases for signature search, users can use the `/listdata` endpoint as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"repositories\": [\n",
      "    {\n",
      "      \"datatype\": \"rank_matrix\",\n",
      "      \"uuid\": \"LINCS chemical perturbagen signatures\"\n",
      "    },\n",
      "    {\n",
      "      \"datatype\": \"rank_matrix\",\n",
      "      \"uuid\": \"LINCS consensus gene (CGS) knockdown signatures\"\n",
      "    },\n",
      "    {\n",
      "      \"datatype\": \"rank_matrix\",\n",
      "      \"uuid\": \"l1000_cp\"\n",
      "    },\n",
      "    {\n",
      "      \"datatype\": \"rank_matrix\",\n",
      "      \"uuid\": \"l1000_aby\"\n",
      "    },\n",
      "    {\n",
      "      \"datatype\": \"rank_matrix\",\n",
      "      \"uuid\": \"l1000_lig\"\n",
      "    },\n",
      "    {\n",
      "      \"datatype\": \"rank_matrix\",\n",
      "      \"uuid\": \"l1000_xpr\"\n",
      "    },\n",
      "    {\n",
      "      \"datatype\": \"rank_matrix\",\n",
      "      \"uuid\": \"LINCS gene overexpression signatures\"\n",
      "    },\n",
      "    {\n",
      "      \"datatype\": \"rank_matrix\",\n",
      "      \"uuid\": \"l1000_oe\"\n",
      "    },\n",
      "    {\n",
      "      \"datatype\": \"rank_matrix\",\n",
      "      \"uuid\": \"l1000_shRNA\"\n",
      "    },\n",
      "    {\n",
      "      \"datatype\": \"rank_matrix\",\n",
      "      \"uuid\": \"l1000_siRNA\"\n",
      "    }\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "\n",
    "API_URL = \"https://maayanlab.cloud/sigcom-lincs/data-api/api/v1/listdata\"\n",
    "\n",
    "res = requests.post(API_URL)\n",
    "databases = res.json()\n",
    "\n",
    "for_enrichment = {\n",
    "    \"up_entities\": [],\n",
    "    \"down_entities\": []\n",
    "}\n",
    "\n",
    "print(json.dumps(databases, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### References <a name=\"ref\"></a>\n",
    "[1] Clark, N. R., Hu, K. S., Feldmann, A. S., Kou, Y., Chen, E. Y., Duan, Q., & Ma’ayan, A. (2014). The characteristic direction: a geometrical approach to identify differentially expressed genes. BMC bioinformatics, 15(1), 1-16. doi: [10.1186/1471-2105-15-79](https://doi.org/10.1186/1471-2105-15-79). PMID: [24650281](https://pubmed.ncbi.nlm.nih.gov/24650281/). PMCID: [PMC4000056](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4000056/). \n",
    "\n",
    "[2] Ritchie, M. E., Phipson, B., Wu, D. I., Hu, Y., Law, C. W., Shi, W., & Smyth, G. K. (2015). limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic acids research, 43(7), e47-e47. doi: [10.1093/nar/gkv007](https://doi.org/10.1093/nar/gkv007). PMID: [25605792](https://pubmed.ncbi.nlm.nih.gov/25605792/). PMCID: [PMC4402510](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4402510/)\n",
    "\n",
    "[3] Subramanian A, Narayan R, Corsello SM, Peck DD, Natoli TE, Lu X, Gould J, Davis JF, Tubelli AA, Asiedu JK, Lahr DL, Hirschman JE, Liu Z, Donahue M, Julian B, Khan M, Wadden D, Smith IC, Lam D, Liberzon A, Toder C, Bagul M, Orzechowski M, Enache OM, Piccioni F, Johnson SA, Lyons NJ, Berger AH, Shamji AF, Brooks AN, Vrcic A, Flynn C, Rosains J, Takeda DY, Hu R, Davison D, Lamb J, Ardlie K, Hogstrom L, Greenside P, Gray NS, Clemons PA, Silver S, Wu X, Zhao WN, Read-Button W, Wu X, Haggarty SJ, Ronco LV, Boehm JS, Schreiber SL, Doench JG, Bittker JA, Root DE, Wong B, Golub TR. (2017). A next generation connectivity map: L1000 platform and the first 1,000,000 profiles. Cell, 171(6), 1437-1452. doi: [10.1016/j.cell.2017.10.049](https://doi.org/10.1016/j.cell.2017.10.049). PMID: [29195078](https://pubmed.ncbi.nlm.nih.gov/29195078). PMCID: [PMC5990023](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5990023). \n",
    "\n",
    "[4] Keenan AB, Torre D, Lachmann A, Leong AK, Wojciechowicz ML, Utti V, Jagodnik KM, Kropiwnicki E, Wang Z, Ma'ayan A. (2019). ChEA3: transcription factor enrichment analysis by orthogonal omics integration. Nucleic acids research, 47(W1), W212-W224. doi: [10.1093/nar/gkz446](https://doi.org/10.1093/nar/gkz446). PMID: [31114921](https://pubmed.ncbi.nlm.nih.gov/31114921); PMCID: [PMC6602523](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6602523).\n",
    "\n",
    "[5] Lachmann A, Torre D, Keenan AB, Jagodnik KM, Lee HJ, Wang L, Silverstein MC, Ma’ayan A. (2018). Massive mining of publicly available RNA-seq data from human and mouse. Nature communications, 9(1), 1-10. doi: [10.1038/s41467-018-03751-6](https://doi.org/10.1038/s41467-018-03751-6). PMID: [29636450](https://pubmed.ncbi.nlm.nih.gov/29636450) PMCID: [PMC5893633](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5893633)\n",
    "\n",
    "[6] Clarke, D. J., Jeon, M., Stein, D. J., Moiseyev, N., Kropiwnicki, E., Dai, C., ... & Ma'ayan, A. (2021). Appyters: Turning Jupyter Notebooks into data-driven web apps. Patterns, 2(3), 100213. doi: [10.1016/j.patter.2021.100213](https://doi.org/10.1016/j.patter.2021.100213). PMID: [33748796](https://pubmed.ncbi.nlm.nih.gov/33748796) PMCID: [PMC7961182](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7961182)\n",
    "\n",
    "[7] Chen EY, Tan CM, Kou Y, Duan Q, Wang Z, Meirelles GV, Clark NR, Ma'ayan A. (2013). Enrichr: interactive and collaborative HTML5 gene list enrichment analysis tool. BMC bioinformatics, 14(1), 1-14. doi: [10.1186/1471-2105-14-128](https://doi.org/10.1186/1471-2105-14-128). PMID: [23586463](https://pubmed.ncbi.nlm.nih.gov/23586463) PMCID: [PMC3637064](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3637064)\n",
    "\n",
    "[8] Kuleshov MV, Jones MR, Rouillard AD, Fernandez NF, Duan Q, Wang Z, Koplev S, Jenkins SL, Jagodnik KM, Lachmann A, McDermott MG, Monteiro CD, Gundersen GW, Ma'ayan A. (2016). Enrichr: a comprehensive gene set enrichment analysis web server 2016 update. Nucleic acids research, 44(W1), W90-W97. doi: [10.1093/nar/gkw377](https://doi.org/10.1093/nar/gkw377). PMID: [27141961](https://pubmed.ncbi.nlm.nih.gov/27141961) PMCID: [PMC4987924](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4987924)\n",
    "\n",
    "[9] Xie Z, Bailey A, Kuleshov MV, Clarke DJB., Evangelista JE, Jenkins SL, Lachmann A, Wojciechowicz ML, Kropiwnicki E, Jagodnik KM, Jeon M, & Ma’ayan A. (2021). Gene set knowledge discovery with Enrichr. Current protocols, 1(3), e90. doi: [10.1002/cpz1.90](https://doi.org/10.1002/cpz1.90). PMID: [33780170](https://pubmed.ncbi.nlm.nih.gov/33780170) PMCID: [PMC8152575](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8152575)\n",
    "\n",
    "[10] Wilkinson MD, Dumontier M, Aalbersberg IJ, Appleton G, Axton M, Baak A, Blomberg N, Boiten JW, da Silva Santos LB, Bourne PE, Bouwman J, Brookes AJ, Clark T, Crosas M, Dillo I, Dumon O, Edmunds S, Evelo CT, Finkers R, Gonzalez-Beltran A, Gray AJ, Groth P, Goble C, Grethe JS, Heringa J, 't Hoen PA, Hooft R, Kuhn T, Kok R, Kok J, Lusher SJ, Martone ME, Mons A, Packer AL, Persson B, Rocca-Serra P, Roos M, van Schaik R, Sansone SA, Schultes E, Sengstag T, Slater T, Strawn G, Swertz MA, Thompson M, van der Lei J, van Mulligen E, Velterop J, Waagmeester A, Wittenburg P, Wolstencroft K, Zhao J, Mons B. (2016). The FAIR Guiding Principles for scientific data management and stewardship. Scientific data, 3(1), 1-9. doi: [10.1038/sdata.2016.18](https://doi.org/10.1038/sdata.2016.18). PMID: [26978244](https://pubmed.ncbi.nlm.nih.gov/26978244) PMCID: [PMC4792175](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4792175) \n",
    "\n",
    "[11] Clarke DJB, Wang L, Jones A, Wojciechowicz ML, Torre D, Jagodnik KM, Jenkins SL, McQuilton P, Flamholz Z, Silverstein MC, Schilder BM, Robasky K, Castillo C, Idaszak R, Ahalt SC, Williams J, Schurer S, Cooper DJ, de Miranda Azevedo R, Klenk JA, Haendel MA, Nedzel J, Avillach P, Shimoyama ME, Harris RM, Gamble M, Poten R, Charbonneau AL, Larkin J, Brown CT, Bonazzi VR, Dumontier MJ, Sansone SA, Ma'ayan A. FAIRshake: toolkit to evaluate the FAIRness of research digital resources. Cell systems, 9(5), 417-421. doi: [10.1016/j.cels.2019.09.011](https://doi.org/10.1016/j.cels.2019.09.011). PMID: [31677972](https://pubmed.ncbi.nlm.nih.gov/31677972) PMCID: [PMC7316196](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7316196) \n",
    "\n",
    "[12] Jun-Yan Zhu*, Taesung Park*, Phillip Isola, and Alexei A. Efros. (2017). Unpaired image-to-image translation using cycle-consistent adversarial networks. In Proceedings of the IEEE international conference on computer vision (pp. 2223-2232). Paper: [arXiv:1703.10593](https://arxiv.org/abs/1703.10593)\n",
    "\n",
    "[13] Zaveri, A., Dastgheib, S., Wu, C., Whetzel, T., Verborgh, R., Avillach, P., Korodi, G., Terryn, R., Jagodnik, K.M., Assis, P., & Dumontier, M. (2017). smartAPI: towards a more intelligent network of web APIs. In European Semantic Web Conference (pp. 154-169). Springer, Cham. doi: [10.1007/978-3-319-58451-5_11](https://doi.org/10.1007/978-3-319-58451-5_11)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter-default",
   "language": "python",
   "name": "jupyter-default"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
